File: dp_labs_and_solutions/dp_labs_and_solutions/dp_java_labs_and_solutions/mediator/lab/Consumer.java
// Consumer.java

public class Consumer
{
	public Consumer(String aName)
	{
		myName = aName;
		myProducers = new DpVector<Producer>();
	}

	public void addProducer(Producer aProducer)
	{
		if (aProducer == null)
		{
			throw new DpException("void addProducer(Producer aProducer) -- aProducer == null.");
		}

		myProducers.add(aProducer);
	}

	public Producer chooseProducer()
	{
		Producer aLowestBidProducer = null;
		int aLowestBid = 0;

		DpIterator<Producer> aIterator = myProducers.makeIterator();
		while (!aIterator.isDone())
		{
			Producer aProducer = aIterator.getCurrent();
			int aBid = aProducer.getBid();
			if (aLowestBid == 0 || aBid < aLowestBid)
			{
				aLowestBid = aBid;
				aLowestBidProducer = aProducer;
			}
			aIterator.advance();
		}

		return aLowestBidProducer;
	}

	public void consume()
	{
		Producer aProducer = chooseProducer();
		if (aProducer == null)
		{
			throw new DpException("void consume() -- aProducer == null.");
		}

		aProducer.produce();
		System.out.println(myName + " consumes.");
	}

	private String myName;
	private DpVector<Producer> myProducers;
}

File: dp_labs_and_solutions/dp_labs_and_solutions/dp_matlab_labs_and_solutions/mediator/lab/Consumer.m
% Consumer.m

classdef Consumer < handle
	properties (GetAccess = public, SetAccess = private)
		Name % string
		Producers % DpVector of Producer
	end

	methods
		% constructor:
		function theConsumer = Consumer(aName)
			theConsumer.Name = aName;
			theConsumer.Producers = DpVector;
		end

		function addProducer(theConsumer, aProducer)
			if isempty(aProducer)
				error('Invalid aProducer.');
			end

			theConsumer.Producers.add(aProducer);
		end

		function result = chooseProducer(theConsumer)
			aLowestBidProducer = Producer.empty;
			aLowestBid = 0;
			aIterator = theConsumer.Producers.makeIterator;
			while ~aIterator.isDone
				aProducer = aIterator.getCurrent;
				aBid = aProducer.getBid;
				if (aLowestBid == 0) || (aBid < aLowestBid)
					aLowestBid = aBid;
					aLowestBidProducer = aProducer;
				end
				aIterator.advance;
			end
			result = aLowestBidProducer;
		end

		function consume(theConsumer)
			aProducer = theConsumer.chooseProducer;
			if isempty(aProducer)
				error('Invalid aProducer.');
			end

			aProducer.produce;
			fprintf('%s consumes.\n', theConsumer.Name);
		end
	end
end
